x/incentive/keeper/update.go

```diff
@@ -6,14 +6,194 @@ import (
+		// program after the previous calculation.
+		prevRemainingTime := (p.StartTime + p.Duration) - prevTime
+		// if remaining time was zero or negative, but program had not been removed
+		// from ongoing programs, using a value of 1 second ensures all its remaining
+		// rewards are distributed
+		if prevRemainingTime < 1 {
+			prevRemainingTime = 1
+		}
+
+		// The portion of a program's remaining rewards,
+		// ranging from 0 to 1, which will be distributed this
+		// block. The max value of 1 means 100% of remaining rewards
+		// will be used, as occurs when a program is ending.
+		programRewardsFraction := sdk.MinDec(
+			sdk.OneDec(),
+			sdk.NewDecFromInt(sdk.NewInt(timeElapsed)).
+				Quo(sdk.NewDec(prevRemainingTime)))
+
+		// each incentive program has only one reward denom
+		rewardDenom := p.RemainingRewards.Denom
+
+		// get this block's rewards (as a token amount) for this incentive program only
+		thisBlockRewards := sdk.NewCoin(
+			rewardDenom,
+			sdk.NewDecFromInt(p.RemainingRewards.Amount).Mul(programRewardsFraction).TruncateInt())
+
+		// get expected increase of bondDenom's rewardAccumulator of reward denom,
+		// by dividing this block's rewards proportionally among bonded uTokens,
+		// and adjusting for the reward accumulator's exponent
+		accumulator := k.getRewardAccumulator(ctx, bondedDenom)
+		accumulatorIncrease := sdk.NewDecFromInt(thisBlockRewards.Amount).
+			Mul(ten.Power(uint64(accumulator.Exponent))).
+			Quo(sdk.NewDecFromInt(bonded.Amount))
+
+		// if accumulator increase is so small it rounds to zero even after power adjustment,
+		// no rewards were distributed
+		if accumulatorIncrease.IsZero() {
+			continue
+		}
+
+		// if nonzero, increase reward accumulator using rewards from this incentive program
+		// and subtract them from that program's remaining rewards
+		accumulator.Rewards = accumulator.Rewards.Add(sdk.NewDecCoinFromDec(rewardDenom, accumulatorIncrease))
+		p.RemainingRewards = p.RemainingRewards.Sub(thisBlockRewards)
+
+		if err := k.setIncentiveProgram(ctx, p, incentive.ProgramStatusOngoing); err != nil {
+			return err
+		}
+		if err := k.setRewardAccumulator(ctx, accumulator); err != nil {
+			return err
+		}
+	}
+
+	// After updates, module's LastRewardTime increases to current block time
+	return k.setLastRewardsTime(ctx, blockTime)
+}
+
+// updatePrograms moves any incentive programs which have reached their end time from Ongoing to Completed,
+// and moves any funded programs which should start from Upcoming to Ongoing. Non-funded programs which
+// would start are moved to completed status instead.
+func (k Keeper) updatePrograms(ctx sdk.Context, blockTime int64) error {
+	ongoingPrograms, err := k.getAllIncentivePrograms(ctx, incentive.ProgramStatusOngoing)
+	if err != nil {
+		return err
+	}
+	for _, op := range ongoingPrograms {
+		// if an ongoing program is ending, move it to completed programs without modifying any fields
+		if op.Duration+op.StartTime >= blockTime {
```

Comment:
Since the order of endblock functions is `updateRewards` -> `updatePrograms`, a program that ends at exact unix time `N` has already released its rewards for time `N` when this line is reached.